Add support for resource conversion plugins in filesystem dock

SaracenOne opened the pull request:
Based on a past discussion for a long overdue UX feature. Makes ResourceConversionPlugins avaliable to the filesystem dock, including allowing multiple files to be batch converted in one go and updating the usage of said resources in active scenes. Keeping in draft for a bit longer since I want to look at revising some of the API naming around detecting viable conversion plugins based purely on class name to avoid having to load the resource beforehand, and possibly look into introducing a warning dialogue explaining how conversion of types is a one-way operation.

That being said, I think the majority of the implementation is fine.

![OLS4nq9wfq](https://user-images.githubusercontent.com/12756047/189425341-30ab6e64-d00a-4b1f-8932-25da476b47bc.png)

SaracenOne committed:
Add support for resource conversion plugins in filesystem dock.

SaracenOne commented:
Decides to revisit this and add the remaining things I wanted to add because, to be frank, the current workflow for converting standard materials to shader materials is kind of atrocious. It now gives you warning when you attempt to convert files from the filesystem dock, and although there are currently no instances where multiple conversion types are avaliable, it will now collapse the conversion choices into a submenu if more than one is present, future.

I'm not happy about the fact that I had to add a new (and fairly reduntant) function to the API. I would have much rathered changes change 'handles' to a string type rather than a resource type, but I realize that doing so would break the API and subsequently break GDExtensions. As such, another function is added to ResourceConverterPlugins to check handling by type name rather than resource, meaning the resources won't have to be loaded when simply right-clicking on them

AThousandShips commented on editor/editor_node.h:
```diff
@@ -669,6 +669,10 @@ class EditorNode : public Node {
 
 	void _resources_changed(const Vector<String> &p_resources);
+
+	void _replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource);
```
```suggestion
	void _replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
```

AThousandShips commented on editor/editor_node.h:
```diff
@@ -669,6 +669,10 @@ class EditorNode : public Node {
 	void _resources_changed(const Vector<String> &p_resources);
+
+	void _replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource);
+	void _replace_resources_in_node(Node *node, const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource);
```
```suggestion
	void _replace_resources_in_node(Node *node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
```

AThousandShips commented on editor/editor_node.h:
```diff
@@ -782,6 +786,8 @@ class EditorNode : public Node {
 	void push_node_item(Node *p_node);
 	void hide_unused_editors(const Object *p_editing_owner = nullptr);
 
+	void replace_resources_in_scenes(const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource);
```
```suggestion
	void replace_resources_in_scenes(const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
```

AThousandShips commented on editor/editor_node.h:
```diff
@@ -928,7 +934,8 @@ class EditorNode : public Node {
 	void remove_resource_conversion_plugin(const Ref<EditorResourceConversionPlugin> &p_plugin);
-	Vector<Ref<EditorResourceConversionPlugin>> find_resource_conversion_plugin(const Ref<Resource> &p_for_resource);
+	Vector<Ref<EditorResourceConversionPlugin>> find_resource_conversion_plugin_for_resource(const Ref<Resource> &p_for_resource);
+	Vector<Ref<EditorResourceConversionPlugin>> find_resource_conversion_plugin_for_type_name(const String p_type);
```
```suggestion
	Vector<Ref<EditorResourceConversionPlugin>> find_resource_conversion_plugin_for_type_name(const String &p_type);
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -6510,11 +6599,25 @@ void EditorNode::remove_resource_conversion_plugin(const Ref<EditorResourceConve
+	return ret;
+}
+
+Vector<Ref<EditorResourceConversionPlugin>> EditorNode::find_resource_conversion_plugin_for_type_name(const String p_type) {
```
```suggestion
Vector<Ref<EditorResourceConversionPlugin>> EditorNode::find_resource_conversion_plugin_for_type_name(const String &p_type) {
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -3208,6 +3290,13 @@ void EditorNode::editor_select(int p_which) {
 	}
 }
 
+void EditorNode::replace_resources_in_scenes(const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource) {
```
```suggestion
void EditorNode::replace_resources_in_scenes(const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -890,6 +890,88 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
+	}
+}
+
+void EditorNode::_replace_resources_in_node(Node *node, const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource) {
```
```suggestion
void EditorNode::_replace_resources_in_node(Node *node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -890,6 +890,88 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
 	}
 }
 
+void EditorNode::_replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> p_source_resources, const Vector<Ref<Resource>> p_target_resource) {
```
```suggestion
void EditorNode::_replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```

AThousandShips commented on editor/filesystem_dock.h:
```diff
@@ -251,6 +257,8 @@ class FileSystemDock : public VBoxContainer {
 	void _file_list_gui_input(Ref<InputEvent> p_event);
 	void _tree_gui_input(Ref<InputEvent> p_event);
 
+	Vector<String> _get_valid_conversions_for_file_paths(Vector<String> p_paths);
```
```suggestion
	Vector<String> _get_valid_conversions_for_file_paths(const Vector<String> &p_paths);
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
 	}
 }
 
+Vector<String> FileSystemDock::_get_valid_conversions_for_file_paths(Vector<String> p_paths) {
```
```suggestion
Vector<String> FileSystemDock::_get_valid_conversions_for_file_paths(const Vector<String> &p_paths) {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1852,6 +1907,50 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+	Vector<String> valid_conversion_targets = _get_valid_conversions_for_file_paths(to_convert);
+
+	Vector<Ref<Resource>> selected_resources;
+	for (String f_path : to_convert) {
```
```suggestion
	for (const String &f_path : to_convert) {
```

AThousandShips commented on editor/plugins/editor_resource_conversion_plugin.h:
```diff
@@ -42,11 +42,16 @@ class EditorResourceConversionPlugin : public RefCounted {
+	// prevent API breakage and have the RMB menu check of valid conversion types without
+	// having to load them. Hopefully we can unify this for Godot 5...
 	virtual bool handles(const Ref<Resource> &p_resource) const;
+	virtual bool handles_type_name(const String p_type) const;
```
```suggestion
	virtual bool handles_type_name(const String &p_type) const;
```

AThousandShips commented on editor/plugins/editor_resource_conversion_plugin.h:
```diff
@@ -42,11 +42,16 @@ class EditorResourceConversionPlugin : public RefCounted {
 
 	GDVIRTUAL0RC(String, _converts_to)
 	GDVIRTUAL1RC(bool, _handles, Ref<Resource>)
+	GDVIRTUAL1RC(bool, _handles_type_name, String)
```
```suggestion
	GDVIRTUAL1RC(bool, _handles_type_name, const String &)
```

AThousandShips commented on editor/plugins/editor_resource_conversion_plugin.cpp:
```diff
@@ -48,6 +49,12 @@ bool EditorResourceConversionPlugin::handles(const Ref<Resource> &p_resource) co
 	return ret;
 }
 
+bool EditorResourceConversionPlugin::handles_type_name(const String p_type) const {
```
```suggestion
bool EditorResourceConversionPlugin::handles_type_name(const String &p_type) const {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+					all_valid_conversion_to_targets = current_valid_conversion_to_targets;
+				} else {
+					// Check existing conversion targets and remove any which are not in the current list.
+					for (String S : all_valid_conversion_to_targets) {
```
```suggestion
					for (const String &S : all_valid_conversion_to_targets) {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -2368,6 +2488,23 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+				ERR_FAIL_COND(selected_conversion_id >= valid_conversion_targets.size());
+
+				to_convert.clear();
+				for (String f_path : p_selected) {
```
```suggestion
				for (const String &f_path : p_selected) {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -3009,6 +3146,47 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, Vector<Str
+						String what = all_valid_conversion_to_targets[i];
+						Ref<Texture2D> icon;
+						if (has_theme_icon(what, SNAME("EditorIcons"))) {
+							icon = get_theme_icon(what, SNAME("EditorIcons"));
```
```suggestion
							icon = get_editor_theme_icon(what);
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -3009,6 +3146,47 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, Vector<Str
+						String what = all_valid_conversion_to_targets[i];
+						Ref<Texture2D> icon;
+						if (has_theme_icon(what, SNAME("EditorIcons"))) {
+							icon = get_theme_icon(what, SNAME("EditorIcons"));
```
```suggestion
							icon = get_editor_theme_icon(what);
```

AThousandShips reviewed (commented):
Dedicated icon syntax

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+						}
+					}
+					// We have no more remaining valid conversions, so return an empty list.
+					if (all_valid_conversion_to_targets.size() == 0) {
```
```suggestion
					if (all_valid_conversion_to_targets.is_empty()) {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+					}
+				}
+
+				if (all_valid_conversion_to_targets.size() == 0) {
```
```suggestion
				if (all_valid_conversion_to_targets.is_empty()) {
```

SaracenOne commented:
Okay thanks, implemented changes recommend by @AThousandShips

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+					all_valid_conversion_to_targets = current_valid_conversion_to_targets;
+				} else {
+					// Check existing conversion targets and remove any which are not in the current list.
+					for (const String S : all_valid_conversion_to_targets) {
```
```suggestion
					for (const String &S : all_valid_conversion_to_targets) {
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -2368,6 +2488,23 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+				ERR_FAIL_COND(selected_conversion_id >= valid_conversion_targets.size());
+
+				to_convert.clear();
+				for (const String f_path : p_selected) {
```
```suggestion
				for (const String &f_path : p_selected) {
```

AThousandShips commented on editor/editor_node.h:
```diff
@@ -669,6 +669,10 @@ class EditorNode : public Node {
 
 	void _resources_changed(const Vector<String> &p_resources);
+
+	void _replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
+	void _replace_resources_in_node(Node *node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
```
```suggestion
	void _replace_resources_in_object_properties(Object *p_obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
	void _replace_resources_in_node(Node *p_node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource);
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -890,6 +890,91 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
+	}
+}
+
+void EditorNode::_replace_resources_in_node(Node *node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```
```suggestion
void EditorNode::_replace_resources_in_node(Node *p_node, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -890,6 +890,91 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
 	}
 }
 
+void EditorNode::_replace_resources_in_object_properties(Object *obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```
```suggestion
void EditorNode::_replace_resources_in_object_properties(Object *p_obj, const Vector<Ref<Resource>> &p_source_resources, const Vector<Ref<Resource>> &p_target_resource) {
```

SaracenOne commented:
Okay, updated now

KoBeWi commented:
I think `handles()` takes a whole object to allow filtering objects by more than only a type (although it's not really used in practice). 

You can use this trick to achieve the same functionality with a low cost:
```C++

	Ref<Resource> temp = ClassDB::instantiate(p_type);
	for (int i = 0; i < resource_conversion_plugins.size(); i++) {
		if (resource_conversion_plugins[i].is_valid() && resource_conversion_plugins[i]->handles(temp)) {
			ret.push_back(resource_conversion_plugins[i]);
		}
	}
```
Instantiating an empty resource is much cheaper than loading it.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+		ERR_FAIL_COND_V(fpath == "", Vector<String>());
+		ERR_FAIL_COND_V(fpath == "res://", Vector<String>());
+		if (!FileAccess::exists(fpath)) {
+			return Vector<String>();
```
Should this return when other paths might be valid?

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+	for (int i = 0; i < p_paths.size(); i++) {
+		String fpath = p_paths[i];
+
+		ERR_FAIL_COND_V(fpath == "", Vector<String>());
```
```suggestion
		ERR_FAIL_COND_V(fpath.is_empty(), Vector<String>());
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+
+			if (conversions.size() != 0) {
+				// Get a list of all potentional conversion-to targets.
+				Vector<String> current_valid_conversion_to_targets;
```
You can change it to HashSet to automatically avoid duplicate entries.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1157,6 +1159,59 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+			if (conversions.size() != 0) {
+				// Get a list of all potentional conversion-to targets.
+				Vector<String> current_valid_conversion_to_targets;
+				for (Ref<EditorResourceConversionPlugin> E : conversions) {
+					String what = E->converts_to();
```
```suggestion
				for (const Ref<EditorResourceConversionPlugin> &E : conversions) {
					const String what = E->converts_to();
```

KoBeWi commented:
I see some faulty logic in `_get_valid_conversions_for_file_paths()`. It returns empty Vector when it encounters any problem, but otherwise it will happily dump every plugin into one vector. The plugins might come from different resource types, so the list may contain conversion entries that only apply for some of the files. If that's the case, why abort when a file has no available conversions? Such files could be simply skipped.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1852,6 +1907,50 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+				ERR_FAIL_COND(!res.is_valid());
+				converted_resources.push_back(converted_res);
+
+				// Clear history for the objects being replaced.
+				EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+				undo_redo->clear_history(true, undo_redo->get_history_id_for_object(res.ptr()));
```
This potentially clears history for the same context multiple times. You could have a HashSet of history ids that need clearing, to avoid repeated clears (it's not a big problem though).

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1852,6 +1907,50 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+		for (Ref<EditorResourceConversionPlugin> E : conversions) {
+			if (E->converts_to() == valid_conversion_targets[selected_conversion_id]) {
+				Ref<Resource> converted_res = E->convert(res);
+				ERR_FAIL_COND(!res.is_valid());
```
```suggestion
				ERR_FAIL_COND(res.is_null());
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1852,6 +1907,50 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+
+		// Overwrite the path and UUID.
+		new_resource->set_path(original_resource->get_path(), true);
+		new_resource->set_scene_unique_id(original_resource->get_scene_unique_id());
```
Scene unique ID is the ID for resources referenced in a scene. It doesn't make sense in a context of stand-alone external resources.

Both UID and path are automatically handled by `ResourceSaver`, so you can remove these 2 lines.

SaracenOne commented:
Sorry I neglected this PR for so long, but okay, I wanted to take another shot of rebasing and attempting to mainline this. I still think its a very handy convinence feature which addresses a couple of needless frictions for the user when it comes to converting standard materials into ShaderMaterials (currently you can only really do it if its already assigned to an inspector property). I've addressed your feedback @KoBeWi and included things like unifying the replace function in the inspector dock, making use of HashSets and made use of your suggestion of using the ClassDB::instantiate method instead.

One thing I do want to question is the issue regarding the behaviour of the `_get_valid_conversions_for_file_paths()`, where it will return empty if *any* error is encountered. I think this might be intended behaviour; the operation is intended to exclusive rather than inclusive, and the part of the function which checks if there are any converters which are not valid across another resource:
```
					// Check existing conversion targets and remove any which are not in the current list.
					for (const String &S : all_valid_conversion_to_targets) {
						if (!current_valid_conversion_to_targets.has(S)) {
							all_valid_conversion_to_targets.erase(S);
						}
					}
```
I might also be missing something, but I believe this behaviour is intended since if the user has multiple resource types selected, it should only exclusively show the conversion targets which are valid for every resource in the selection.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1178,6 +1180,54 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
 
+HashSet<String> FileSystemDock::_get_valid_conversions_for_file_paths(const Vector<String> &p_paths) {
+	HashSet<String> all_valid_conversion_to_targets;
+	for (int i = 0; i < p_paths.size(); i++) {
+		String fpath = p_paths[i];
+
```
```suggestion
	for (const String &fpath : p_paths) {
```

KoBeWi commented on editor/filesystem_dock.h:
```diff
@@ -140,6 +140,7 @@ class FileSystemDock : public VBoxContainer {
 		FILE_NEW_FOLDER,
 		FILE_NEW_SCRIPT,
 		FILE_NEW_SCENE,
+		CONVERT_BASE_ID = 1000
```
```suggestion
		CONVERT_BASE_ID = 1000,
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -2116,6 +2216,27 @@ void FileSystemDock::_file_list_rmb_option(int p_option) {
+	// calling from the file_list_rmb menu or the _tree_rmb option.
+	Vector<String> selected;
+
+	if (files->has_focus()) {
+		Vector<int> files_selected_ids = files->get_selected_items();
+		for (int i = 0; i < files_selected_ids.size(); i++) {
+			selected.push_back(files->get_item_metadata(files_selected_ids[i]));
+		}
+	} else {
+		TreeItem *tree_selected = tree->get_root();
+		tree_selected = tree->get_next_selected(tree_selected);
+		while (tree_selected) {
+			selected.push_back(tree_selected->get_metadata(0));
+			tree_selected = tree->get_next_selected(tree_selected);
+		}
+	}
```
You could call `_file_list_rmb_option()` or `_tree_rmb_option()` instead of repeating their code.

KoBeWi commented:
Found a bug: if there are multiple conversion options, the first one in the list will always be used regardless of what you select.
I wrote some dumb custom plugin for testing:
```GDScript
@tool
extends EditorPlugin

func _enter_tree() -> void:
	add_resource_conversion_plugin(Conver.new())

class Conver extends EditorResourceConversionPlugin:
	func _handles(resource: Resource) -> bool:
		return resource is Material
	
	func _converts_to() -> String:
		return "GDScript"
	
	func _convert(resource: Resource) -> Resource:
		var code := """extends Node

func _ready():
	print(%s.new())
""" % resource.get_class()
		
		var gd := GDScript.new()
		gd.source_code = code
		return gd
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -1898,6 +1948,56 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+		new_resource->set_path(original_resource->get_path(), true);
+
+		ResourceSaver::save(converted_resources[i], to_convert[i]);
+	};
```
```suggestion
	}
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1178,6 +1180,54 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+			}
+		} else {
+			// One of the resources was imported, so return an empty list.
+			return HashSet<String>();
```
This can be an early return instead.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1178,6 +1180,54 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+				}
+			} else {
+				// This resource can't convert to anything, so return an empty list.
+				return HashSet<String>();
```
Also can be early return.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1178,6 +1180,54 @@ void FileSystemDock::_update_file_list(bool p_keep_selection) {
+					}
+					// We have no more remaining valid conversions, so return an empty list.
+					if (all_valid_conversion_to_targets.is_empty()) {
+						return HashSet<String>();
```
```suggestion
						return all_valid_conversion_to_targets;
```
It's empty anyway.

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -6417,12 +6509,31 @@ void EditorNode::remove_resource_conversion_plugin(const Ref<EditorResourceConve
+	if (ClassDB::can_instantiate(p_type)) {
+		Ref<Resource> temp = Object::cast_to<Resource>(ClassDB::instantiate(p_type));
+		if (temp.is_valid()) {
+			for (int i = 0; i < resource_conversion_plugins.size(); i++) {
+				if (resource_conversion_plugins[i].is_valid() && resource_conversion_plugins[i]->handles(temp)) {
+					ret.push_back(resource_conversion_plugins[i]);
```
```suggestion
			for (Ref<EditorResourceConversionPlugin> resource_conversion_plugin : resource_conversion_plugins) {
				if (resource_conversion_plugin.is_valid() && resource_conversion_plugin->handles(temp)) {
					ret.push_back(resource_conversion_plugin);
```

AThousandShips commented on editor/editor_node.cpp:
```diff
@@ -6417,12 +6509,31 @@ void EditorNode::remove_resource_conversion_plugin(const Ref<EditorResourceConve
+	Vector<Ref<EditorResourceConversionPlugin>> ret;
+
+	if (p_for_resource.is_valid()) {
+		for (int i = 0; i < resource_conversion_plugins.size(); i++) {
+			if (resource_conversion_plugins[i].is_valid() && resource_conversion_plugins[i]->handles(p_for_resource)) {
+				ret.push_back(resource_conversion_plugins[i]);
```
```suggestion
		for (Ref<EditorResourceConversionPlugin> resource_conversion_plugin : resource_conversion_plugins) {
			if (resource_conversion_plugin.is_valid() && resource_conversion_plugin->handles(p_for_resource)) {
				ret.push_back(resource_conversion_plugin);
```

AThousandShips commented on editor/filesystem_dock.cpp:
```diff
@@ -2116,6 +2216,27 @@ void FileSystemDock::_file_list_rmb_option(int p_option) {
+
+	if (files->has_focus()) {
+		Vector<int> files_selected_ids = files->get_selected_items();
+		for (int i = 0; i < files_selected_ids.size(); i++) {
+			selected.push_back(files->get_item_metadata(files_selected_ids[i]));
```
```suggestion
		for (int files_selected_id : files_selected_ids) {
			selected.push_back(files->get_item_metadata(files_selected_id));
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3238,6 +3382,48 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+						new_menu->add_icon_item(icon, vformat(TTR("%s"), E), CONVERT_BASE_ID + relative_id);
+					}
+
+					relative_id++;
```
Well this is the reason for always using the first option.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3238,6 +3382,48 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+							icon = get_theme_icon(E, SNAME("Resource"));
+						}
+
+						p_popup->add_icon_item(icon, vformat(TTR("Convert to %s"), E), CONVERT_BASE_ID + relative_id);
```
```suggestion
						p_popup->add_icon_item(icon, vformat(TTR("Convert to %s"), E), CONVERT_BASE_ID);
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -2556,6 +2677,29 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+		default: {
+			// Resource conversion commands:
+			if (p_option >= CONVERT_BASE_ID) {
+				HashSet<String> valid_conversion_targets = _get_valid_conversions_for_file_paths(p_selected);
```
You are calling it 3 times for the same file. Can it be cached when menu is created?

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -2556,6 +2677,29 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+				int conversion_id = 0;
+				for (const String &E : valid_conversion_targets) {
+					if (conversion_id == selected_conversion_id) {
+						conversion_dialog->set_text(vformat(TTR("Do you wish to convert these files to %s? (This operation cannot be undone!)"), E));
```
The conversion dialog has no other text, so you can set it in the constructor instead of every time.

SaracenOne replied:
I'm a bit confused about this since the text set here is formatted to include the name of the resource type you're converting to. Can you clarify what you mean?

KoBeWi replied:
I mean that I missed vformat, sorry.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3238,6 +3382,48 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+			HashSet<String> all_valid_conversion_to_targets = _get_valid_conversions_for_file_paths(p_paths);
+
+			int relative_id = 0;
+			if (all_valid_conversion_to_targets.size()) {
```
```suggestion
			if (!all_valid_conversion_to_targets.is_empty()) {
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3238,6 +3382,48 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+						if (has_theme_icon(E, SNAME("EditorIcons"))) {
+							icon = get_editor_theme_icon(E);
+						} else {
+							icon = get_theme_icon(E, SNAME("Resource"));
```
Doesn't seem to be working for some reason 🤔
![image](https://github.com/user-attachments/assets/36046561-c840-4edb-bedc-221ed9e94e9f)

AThousandShips replied:
What would the theme item of `E` in the `Resource` theme class mean?

KoBeWi replied:
Ah yeah, it should be `get_editor_theme_icon(SNAME("Resource"));`

KoBeWi replied:
Still wrong.

KoBeWi replied:
I realized there is no icon for Resource, so it would be wrong either way.

SaracenOne commented:
Okay @KoBeWi and @AThousandShips, I think that should address most of the issues raised. One I'm not sure about is @KoBeWi's point about setting conversion dialog text in the constructor.

KoBeWi commented on editor/filesystem_dock.h:
```diff
@@ -226,6 +229,9 @@ class FileSystemDock : public VBoxContainer {
 	bool to_move_or_copy = false;
 
+	Vector<String> to_convert;
+	int selected_conversion_id;
```
```suggestion
	int selected_conversion_id = 0;
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3256,6 +3379,51 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+			if (!cached_valid_conversion_targets.is_empty()) {
+				p_popup->add_separator();
+
+				// If we have more the one type we can convert into, collapse it into a submenu.
```
```suggestion
				// If we have more than one type we can convert into, collapse it into a submenu.
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3256,6 +3379,51 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+
+				if (cached_valid_conversion_targets.size() > CONVERSION_SUBMENU_THRESHOLD) {
+					PopupMenu *new_menu = memnew(PopupMenu);
+					new_menu->set_name("Convert to");
```
You can remove the name and use `add_submenu_node_item()` instead.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -6350,12 +6442,31 @@ void EditorNode::remove_resource_conversion_plugin(const Ref<EditorResourceConve
+Vector<Ref<EditorResourceConversionPlugin>> EditorNode::find_resource_conversion_plugin_for_resource(const Ref<Resource> &p_for_resource) {
+	Vector<Ref<EditorResourceConversionPlugin>> ret;
+
+	if (p_for_resource.is_valid()) {
```
You can make early return instead.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1908,6 +1951,54 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+
+	// Clear history for the objects being replaced.
+	for (Ref<Resource> res : resources_to_erase_history_for) {
+		EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
```
You can put `undo_redo` outside the loop.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1908,6 +1951,54 @@ void FileSystemDock::_overwrite_dialog_action(bool p_overwrite) {
+
+		// Overwrite the path.
+		new_resource->set_path(original_resource->get_path(), true);
+
+		ResourceSaver::save(converted_resources[i], to_convert[i]);
```
Can be simplified to
```suggestion
		ResourceSaver::save(new_resource);
```
(it automatically uses the path you set above)

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -2568,9 +2669,31 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+			// Resource conversion commands:
+			if (p_option >= CONVERT_BASE_ID) {
+				selected_conversion_id = p_option - CONVERT_BASE_ID;
+				ERR_FAIL_COND(selected_conversion_id >= (int)cached_valid_conversion_targets.size());
```
```suggestion
				ERR_FAIL_INDEX(selected_conversion_id, (int)cached_valid_conversion_targets.size());
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3256,6 +3379,51 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
+						relative_id++;
+					}
+				} else {
+					for (const String &E : cached_valid_conversion_targets) {
```
You could unify this loop with the one above by using `p_popup` in place of `new_menu` (probably rename the variable then).

KoBeWi reviewed (approved):
The feature works correctly. I left some more style comments, but the code is good enough to be fixed in a follow-up.

SaracenOne commented:
Updated again! @KoBeWi I wasn't sure if that was what you meant by an 'early return', but can you confirm that the change I made was what you were suggesting?

akien-mga commented:
Thanks!
